File: m26-cs/M26/Joakimsoftware.M26/src/Speed.cs
using System;

// Instances of this class represent the calculated Speed for a given
// Distance and ElapsedTime pair. 
// Chris Joakim, 2021/07/19

namespace Joakimsoftware.M26 {
    public class Speed {

        public Distance distance { get; }
        public ElapsedTime elapsedTime { get; }

        public Speed(Distance dist, ElapsedTime time) {

            distance = dist;
            elapsedTime = time;
        }

        public double mph() {

            return distance.asMiles() / elapsedTime.hours();
        }

        public double kph() {

            return distance.asKilometers() / elapsedTime.hours();
        }

        public double yph() {

            return distance.asYards() / elapsedTime.hours();
        }

        public double secondsPerMile() {

            return elapsedTime.secs / distance.asMiles();
        }

        public string pacePerMile() {

            double spm = secondsPerMile();
            int mm = (int) Math.Floor(spm / 60.0f);
            double ss = (spm - (mm * 60.0f));
            if (ss > 59.999999) {
                ss = 59.999999;
            }
            string mms = mm.ToString("00");
            string sss = ss.ToString("00.00");
            return $"{mms}:{sss}";
        }

        public Speed ageGraded(Age eventAge, Age gradedAge) {

            double agFactor = eventAge.maxPulse() / gradedAge.maxPulse();
            double gradedSecs = ((double) elapsedTime.secs) * agFactor;
            ElapsedTime gradedEt = new ElapsedTime(gradedSecs);
            return new Speed(distance, gradedEt);
        }

        public ElapsedTime projectedTime(Distance anotherDistance, string formula = Constants.SpeedFormulaSimple) {

            if (Constants.SpeedFormulaRiegel.Equals(formula.Trim().ToLower())) {
                double s1 = elapsedTime.secs;
                double m1 = distance.asMiles();
                double m2 = anotherDistance.asMiles();
                double s2 = s1 * Math.Pow((m2/m1), (double) 1.06);
                return new ElapsedTime(s2);
            }
            else {
                double secs = secondsPerMile() * anotherDistance.asMiles();
                return new ElapsedTime(secs);
            }
        }
    }
}

File: m26-py/m26/speed.py
__author__ = 'cjoakim'

import math

from .elapsed_time import ElapsedTime


class Speed(object):

    def __init__(self, d, et):
        self.dist = d    # an instance of Distance
        self.etime = et  # an instance of ElapsedTime

    def mph(self):
        return self.dist.as_miles() / self.etime.hours()

    def kph(self):
        return self.dist.as_kilometers() / self.etime.hours()

    def yph(self):
        return self.dist.as_yards() / self.etime.hours()

    def pace_per_mile(self):
        spm = self.seconds_per_mile()
        mm = math.floor(spm / 60.0)
        ss = spm - (mm * 60.0)

        if ss < 10:
            ss = "0{0}".format(ss)
        else:
            ss = "{0}".format(ss)

        if len(ss) > 5:
            ss = ss[0:5]

        return "{0}:{1}".format(mm, ss)

    def seconds_per_mile(self):
        return float(self.etime.secs / self.dist.as_miles())

    def projected_time(self, another_distance, algorithm='simple'):
        if algorithm is 'riegel':
            t1 = float(self.etime.secs)
            d1 = self.dist.as_miles()
            d2 = another_distance.as_miles()
            t2 = t1 * math.pow(float(d2 / d1), float(1.06))
            et = ElapsedTime(t2)
            return et.as_hhmmss()
        else:
            secs = float(self.seconds_per_mile() * another_distance.as_miles())
            et = ElapsedTime(secs)
            return et.as_hhmmss()

    def age_graded(self, event_age, graded_age):
        ag_factor = event_age.max_pulse() / graded_age.max_pulse()
        graded_secs = float((self.etime.secs)) * float(ag_factor)
        graded_et = ElapsedTime(graded_secs)
        return Speed(self.dist, graded_et)

    def __str__(self):
        template = "<Speed dist:{0} etime:{1}>"
        return template.format(self.dist, self.etime)
